pkg/watcher/certificate.go

```diff
@@ -0,0 +1,263 @@
+package watcher
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/go-logr/logr"
+	"github.com/kyma-project/lifecycle-manager/api/v1alpha1"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	v1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	metav1 "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	apimachinerymetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8slabels "k8s.io/apimachinery/pkg/labels"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// private key will only be generated if one does not already exist in the target `spec.secretName`.
+	privateKeyRotationPolicy = "Never"
+
+	DomainAnnotation  = "skr-domain"
+	CertificateSuffix = "-watcher-certificate"
+
+	caCertKey        = "ca.crt"
+	tlsCertKey       = "tls.crt"
+	tlsPrivateKeyKey = "tls.key"
+)
+
+var LabelSet = k8slabels.Set{ //nolint:gochecknoglobals
+	v1alpha1.PurposeLabel: v1alpha1.CertManager,
+	v1alpha1.ManagedBy:    v1alpha1.OperatorName,
+}
+
+type SubjectAltName struct {
+	DNSNames       []string
+	IPAddresses    []string
+	URIs           []string
+	EmailAddresses []string
+}
+
+type Certificate struct {
+	kcpClient       client.Client
+	kyma            *v1alpha1.Kyma
+	certificateName string
+	secretName      string
+	logger          logr.Logger
+}
+
+type CertificateSecret struct {
+	CACrt           string
+	TLSCrt          string
+	TLSKey          string
+	ResourceVersion string
+}
+
+// NewCertificate returns a new Certificate, which can be used for creating a cert-manager Certificate.
+func NewCertificate(kcpClient client.Client, kyma *v1alpha1.Kyma) (*Certificate, error) {
+	if kcpClient == nil || kyma == nil {
```

Comment:
is this ever a valid scenario? Kyma and KCP Client are always present if you're in the reconcile loop

Reply:
You are right, they should always be not nil.